File: finder4j-backtrack/src/test/java/com/github/k0zka/finder4j/backtrack/BacktrackServiceTest.java
package com.github.k0zka.finder4j.backtrack;

import java.util.Arrays;
import java.util.Collections;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.runners.MockitoJUnitRunner;

@RunWith(MockitoJUnitRunner.class)
public class BacktrackServiceTest {

	@Mock
	State state;

	@Mock
	State finalState;

	@Mock
	Step<State> step1;

	@Mock
	Step<State> step2;

	@Mock
	StepFactory<Step<State>, State> factory;

	@Mock
	SolutionListener<State, Step<State>> listener;

	@Mock
	TerminationStrategy<State> terminationStrategy;

	BacktrackService service;

	@Before
	public void startup() {
		// for testing, enforce having 4 workers even if the number of actual
		// CPUs is less
		service = new BacktrackService(4);
	}

	@After
	public void cleanup() {
		service.stop();
	}

	@Test
	public void backtrackParallel() {
		Mockito.when(factory.produce(Mockito.eq(state))).thenReturn(
				Arrays.asList(step1, step2));
		Mockito.when(factory.produce(Mockito.eq(finalState))).thenReturn(
				Collections.<Step<State>> emptyList());
		Mockito.when(step1.take(Mockito.eq(state))).thenReturn(finalState);
		Mockito.when(step2.take(Mockito.eq(state))).thenReturn(finalState);
		Mockito.when(finalState.isComplete()).thenReturn(true);

		service.backtrack(state, factory, terminationStrategy, listener);

		Mockito.verify(listener, Mockito.atLeast(1)).onSolution(
				Mockito.eq(finalState));
	}

}

File: finder4j-backtrack/src/test/kotlin/com/github/k0zka/finder4j/backtrack/BacktrackServiceTest.kt
package com.github.k0zka.finder4j.backtrack

import com.nhaarman.mockito_kotlin.atLeast
import com.nhaarman.mockito_kotlin.eq
import com.nhaarman.mockito_kotlin.mock
import com.nhaarman.mockito_kotlin.verify
import com.nhaarman.mockito_kotlin.whenever
import org.junit.After
import org.junit.Before
import org.junit.Test

class BacktrackServiceTest {

	private var state: State = mock()

	private var finalState: State = mock()

	private var step1: Step<State> = mock()

	var step2: Step<State> = mock()

	var factory: StepFactory<Step<State>, State> = mock()

	var listener: SolutionListener<State, Step<State>> = mock()

	var terminationStrategy: TerminationStrategy<State> = mock()

	var service: BacktrackService? = null

	@Before
	fun startup() {
		// for testing, enforce having 4 workers even if the number of actual
		// CPUs is less
		service = BacktrackService(4)
	}

	@After
	fun cleanup() {
		service?.stop()
	}

	@Test
	fun backtrackParallel() {
		whenever(factory.produce(eq(state))).thenReturn(
				listOf(step1, step2))
		whenever(factory.produce(eq(finalState))).thenReturn(
				listOf())
		whenever(step1.take(eq(state))).thenReturn(finalState)
		whenever(step2.take(eq(state))).thenReturn(finalState)
		whenever(finalState.complete).thenReturn(true)

		service!!.backtrack(state, factory, terminationStrategy, listener)

		verify(listener, atLeast(1)).onSolution(
				eq(finalState))
	}

}
